src/data/setoid/basic.lean

```diff
@@ -359,4 +365,25 @@ def correspondence (r : setoid α) : {s // r ≤ s} ≃o setoid (quotient r) :=
       t.1.trans' (t.1.symm' $ t.2 $ eq_rel.1 hx) $ t.1.trans' ht $ t.2 $ eq_rel.1 hy,
       λ h x y hs, let ⟨a, b, hx, hy, Hs⟩ := hs in ⟨a, b, hx, hy, h Hs⟩⟩ }
 
+lemma subsingleton_quotient_iff_eq_top  {s : setoid α} :
+  subsingleton (quotient s) ↔ s = ⊤ :=
+begin
+  simp only [subsingleton_iff, eq_top_iff, setoid.le_def, setoid.top_def, top_apply, forall_const],
+  refine (surjective_quotient_mk _).forall.trans (forall_congr $ λ a, _),
+  refine (surjective_quotient_mk _).forall.trans (forall_congr $ λ b, _),
+  exact quotient.eq',
+end
+
+@[simp]
+lemma top_rel_eq_triv_rel : (⊤ : α → α → Prop) = λ _ _, true := by {refl}
+
+lemma subsingleton_quot_iff_forall (r: α → α → Prop) : subsingleton (quot r) ↔ eqv_gen r = ⊤ :=
+begin
+  simp only [subsingleton_iff, eq_top_iff, setoid.le_def, setoid.top_def, top_apply, forall_const],
```

Comment:
The setoid lemmas are now irrelevant, you're not using setoids!
```suggestion
  simp only [subsingleton_iff, eq_top_iff, pi.le_def, top_apply, forall_const],
```

Reply:
hmm that broke the proof, the best I could do was ``simp only [subsingleton_iff, eq_top_iff, setoid.le_def, top_apply, forall_const],``